SIGSEV: Glob not applied to shape after Update.

I have two shapes two which i am trying to apply Styles with Glob. This was working perfectly fine before update.

**Current D2 Version**: 0.6.1



Style is stored in /lib/user.d2
```
icon: https://icons.terrastruct.com/essentials%2F359-users.svg
```

## Without Glob

```
production: {
  # ####  Entities ########
  userWebsite
  userMobile

  ## Globs
  userWebsite: @lib/user
  userMobile: @lib/user
}
```

![image](https://github.com/terrastruct/d2/assets/409973/270f448b-e14c-48cb-a470-21f57803892b)

## With Glob

```
production: {
  # ####  Entities ########
  userWebsite
  userMobile

  ## Globs
  user*: @lib/user
}
```

<details><summary>signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x55de6cb4d82f</summary>
<p>
d2 compose.d2
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x55de6cb4d82f]

goroutine 36 [running]:
oss.terrastruct.com/d2/d2ir.(*Map).ensureField.func3()
        /build/d2/src/d2/d2ir/d2ir.go:896 +0x8f
oss.terrastruct.com/d2/d2ir.(*Map).ensureField(0xc0005993e0, 0x0, 0xc000599380, 0xc0003c2a50, 0x1, 0x0, 0xc0000dc1a8, 0xc0000daef0)
        /build/d2/src/d2/d2ir/d2ir.go:915 +0x1515
oss.terrastruct.com/d2/d2ir.(*Map).EnsureField(0xc0003c2a20?, 0xc000599380, 0xc0003c2a50, 0x40?, 0xc0000dc1a8)
        /build/d2/src/d2/d2ir/d2ir.go:721 +0x345
oss.terrastruct.com/d2/d2ir.(*compiler).compileField(0xc0000dc1a8, 0x0?, 0xc0000db058?, 0xc0003c2a50)
        /build/d2/src/d2/d2ir/compile.go:567 +0x4e
oss.terrastruct.com/d2/d2ir.(*compiler).compileKey(0xc0000dc1a8, 0xc0003c2a50)
        /build/d2/src/d2/d2ir/compile.go:547 +0xd8
oss.terrastruct.com/d2/d2ir.(*compiler).compileMap(0xc0000dc1a8, 0xc0005993e0, 0xc000599320, 0xc000599320)
        /build/d2/src/d2/d2ir/compile.go:447 +0x78c
oss.terrastruct.com/d2/d2ir.(*compiler).__import(0xc0000dc1a8, 0xc000508000?)
        /build/d2/src/d2/d2ir/import.go:113 +0x525
oss.terrastruct.com/d2/d2ir.(*compiler)._import(0xc0000dc1a8, 0x0?)
        /build/d2/src/d2/d2ir/import.go:60 +0x2c
oss.terrastruct.com/d2/d2ir.(*compiler)._compileField(0xc0000dc1a8, 0xc00048f630, 0xc0003c2900)
        /build/d2/src/d2/d2ir/compile.go:749 +0x786
oss.terrastruct.com/d2/d2ir.(*compiler).compileField(0xc0000dc1a8, 0xc0003c2900?, 0xc0000db8b0?, 0xc0003c2900)
        /build/d2/src/d2/d2ir/compile.go:574 +0x199
oss.terrastruct.com/d2/d2ir.(*compiler).compileKey(0xc0000dc1a8, 0xc0003c2900)
        /build/d2/src/d2/d2ir/compile.go:547 +0xd8
oss.terrastruct.com/d2/d2ir.(*compiler).compileMap(0xc0000dc1a8, 0xc000599140, 0xc000598ea0, 0xc000598de0)
        /build/d2/src/d2/d2ir/compile.go:447 +0x78c
oss.terrastruct.com/d2/d2ir.(*compiler)._compileField(0xc0000dc1a8, 0xc00048f5e0, 0xc0003c26f0)
        /build/d2/src/d2/d2ir/compile.go:742 +0x1166
oss.terrastruct.com/d2/d2ir.(*compiler).compileField(0xc0000dc1a8, 0x0?, 0xc0000dbeb8?, 0xc0003c26f0)
        /build/d2/src/d2/d2ir/compile.go:574 +0x199
oss.terrastruct.com/d2/d2ir.(*compiler).compileKey(0xc0000dc1a8, 0xc0003c26f0)
        /build/d2/src/d2/d2ir/compile.go:547 +0xd8
oss.terrastruct.com/d2/d2ir.(*compiler).compileMap(0xc0000dc1a8, 0xc000599080, 0xc000598de0, 0xc000598de0)
        /build/d2/src/d2/d2ir/compile.go:447 +0x78c
oss.terrastruct.com/d2/d2ir.Compile(0xc000598de0, 0xc00056f4a0?)
        /build/d2/src/d2/d2ir/compile.go:77 +0x3e7
oss.terrastruct.com/d2/d2compiler.Compile({0xc000042460?, 0x55de6c34e4ae?}, {0x55de6dc5ddc8?, 0xc00056f4a0?}, 0xc00048f450?)
        /build/d2/src/d2/d2compiler/compile.go:42 +0xc6
oss.terrastruct.com/d2/d2lib.Compile({0x55de6dc66d80, 0xc0004d0fc0}, {0xc0000462a0, 0x66}, 0x0?, 0x0?)
        /build/d2/src/d2/d2lib/d2.go:50 +0x155
oss.terrastruct.com/d2/d2cli.compile({0x55de6dc66d80?, 0xc0004d0fc0}, 0xc0004d0f50, {0xc00056f380, 0x2, 0x2}, 0x0, {0x0, 0x0, 0x0, ...}, ...)
        /build/d2/src/d2/d2cli/main.go:394 +0x2fe
oss.terrastruct.com/d2/d2cli.Run({0x55de6dc66d10?, 0xc00048f3b0?}, 0xc0004d0f50)
        /build/d2/src/d2/d2cli/main.go:334 +0x275f
oss.terrastruct.com/util-go/xmain.(*State).Main.func1()
        /build/d2/src/pkg/mod/oss.terrastruct.com/util-go@v0.0.0-20230604222829-11c3c60fec14/xmain/xmain.go:105 +0x66
created by oss.terrastruct.com/util-go/xmain.(*State).Main in goroutine 1
        /build/d2/src/pkg/mod/oss.terrastruct.com/util-go@v0.0.0-20230604222829-11c3c60fec14/xmain/xmain.go:103 +0x139
</details>


</p>
</details>

--- a/ci/release/changelogs/next.md
+++ b/ci/release/changelogs/next.md
@@ -33,3 +33,4 @@
 - Fixes incorrect grid cell positioning when the grid has a shape set and fixes content sometimes escaping circle shapes. [#1734](https://github.com/terrastruct/d2/pull/1734)
 - Fixes content sometimes escaping cloud shapes. [#1736](https://github.com/terrastruct/d2/pull/1736)
 - Fixes panic using a glob filter (e.g. `&a`) outside globs. [#1748](https://github.com/terrastruct/d2/pull/1748)
+- Fixes glob keys with import values (e.g. `user*: @lib/user`). [#1755](https://github.com/terrastruct/d2/pull/1755)

--- a/d2ir/compile.go
+++ b/d2ir/compile.go
@@ -419,8 +419,10 @@ func (c *compiler) ampersandFilterMap(dst *Map, ast, scopeAST *d2ast.Map) bool {
 func (c *compiler) compileMap(dst *Map, ast, scopeAST *d2ast.Map) {
 	var globs []*globContext
 	if len(c.globContextStack) > 0 {
-		previousGlobs := c.globContextStack[len(c.globContextStack)-1]
-		if NodeBoardKind(dst) == BoardLayer {
+		previousGlobs := c.globContexts()
+		// A root layer with existing glob context stack implies it's an import
+		// In which case, the previous globs should be inherited (the else block)
+		if NodeBoardKind(dst) == BoardLayer && !dst.Root() {
 			for _, g := range previousGlobs {
 				if g.refctx.Key.HasTripleGlob() {
 					globs = append(globs, g.prefixed(dst))

--- a/d2ir/import_test.go
+++ b/d2ir/import_test.go
@@ -184,6 +184,21 @@ label: meow`,
 				assertQuery(t, m, 0, 0, 1, "hi")
 			},
 		},
+		{
+			name: "pattern-value",
+			run: func(t testing.TB) {
+				_, err := compileFS(t, "index.d2", map[string]string{
+					"index.d2": `userWebsite
+userMobile
+
+user*: @x
+`,
+					"x.d2": `shape: person
+label: meow`,
+				})
+				assert.Success(t, err)
+			},
+		},
 	}
 
 	runa(t, tca)

--- a/testdata/d2ir/TestCompile/imports/pattern-value.exp.json
+++ b/testdata/d2ir/TestCompile/imports/pattern-value.exp.json
@@ -1,0 +1,447 @@
+{
+  "fields": [
+    {
+      "name": "userWebsite",
+      "composite": {
+        "fields": [
+          {
+            "name": "shape",
+            "primary": {
+              "value": {
+                "range": "x.d2,0:7:7-0:13:13",
+                "value": [
+                  {
+                    "string": "person",
+                    "raw_string": "person"
+                  }
+                ]
+              }
+            },
+            "references": [
+              {
+                "string": {
+                  "range": "x.d2,0:0:0-0:5:5",
+                  "value": [
+                    {
+                      "string": "shape",
+                      "raw_string": "shape"
+                    }
+                  ]
+                },
+                "key_path": {
+                  "range": "x.d2,0:0:0-0:5:5",
+                  "path": [
+                    {
+                      "unquoted_string": {
+                        "range": "x.d2,0:0:0-0:5:5",
+                        "value": [
+                          {
+                            "string": "shape",
+                            "raw_string": "shape"
+                          }
+                        ]
+                      }
+                    }
+                  ]
+                },
+                "context": {
+                  "edge": null,
+                  "key": {
+                    "range": "x.d2,0:0:0-0:13:13",
+                    "key": {
+                      "range": "x.d2,0:0:0-0:5:5",
+                      "path": [
+                        {
+                          "unquoted_string": {
+                            "range": "x.d2,0:0:0-0:5:5",
+                            "value": [
+                              {
+                                "string": "shape",
+                                "raw_string": "shape"
+                              }
+                            ]
+                          }
+                        }
+                      ]
+                    },
+                    "primary": {},
+                    "value": {
+                      "unquoted_string": {
+                        "range": "x.d2,0:7:7-0:13:13",
+                        "value": [
+                          {
+                            "string": "person",
+                            "raw_string": "person"
+                          }
+                        ]
+                      }
+                    }
+                  }
+                },
+                "due_to_glob": true,
+                "due_to_lazy_glob": false
+              }
+            ]
+          },
+          {
+            "name": "label",
+            "primary": {
+              "value": {
+                "range": "x.d2,1:7:21-1:11:25",
+                "value": [
+                  {
+                    "string": "meow",
+                    "raw_string": "meow"
+                  }
+                ]
+              }
+            },
+            "references": [
+              {
+                "string": {
+                  "range": "x.d2,1:0:14-1:5:19",
+                  "value": [
+                    {
+                      "string": "label",
+                      "raw_string": "label"
+                    }
+                  ]
+                },
+                "key_path": {
+                  "range": "x.d2,1:0:14-1:5:19",
+                  "path": [
+                    {
+                      "unquoted_string": {
+                        "range": "x.d2,1:0:14-1:5:19",
+                        "value": [
+                          {
+                            "string": "label",
+                            "raw_string": "label"
+                          }
+                        ]
+                      }
+                    }
+                  ]
+                },
+                "context": {
+                  "edge": null,
+                  "key": {
+                    "range": "x.d2,1:0:14-1:11:25",
+                    "key": {
+                      "range": "x.d2,1:0:14-1:5:19",
+                      "path": [
+                        {
+                          "unquoted_string": {
+                            "range": "x.d2,1:0:14-1:5:19",
+                            "value": [
+                              {
+                                "string": "label",
+                                "raw_string": "label"
+                              }
+                            ]
+                          }
+                        }
+                      ]
+                    },
+                    "primary": {},
+                    "value": {
+                      "unquoted_string": {
+                        "range": "x.d2,1:7:21-1:11:25",
+                        "value": [
+                          {
+                            "string": "meow",
+                            "raw_string": "meow"
+                          }
+                        ]
+                      }
+                    }
+                  }
+                },
+                "due_to_glob": true,
+                "due_to_lazy_glob": false
+              }
+            ]
+          }
+        ],
+        "edges": null
+      },
+      "references": [
+        {
+          "string": {
+            "range": "index.d2,0:0:0-0:11:11",
+            "value": [
+              {
+                "string": "userWebsite",
+                "raw_string": "userWebsite"
+              }
+            ]
+          },
+          "key_path": {
+            "range": "index.d2,0:0:0-0:11:11",
+            "path": [
+              {
+                "unquoted_string": {
+                  "range": "index.d2,0:0:0-0:11:11",
+                  "value": [
+                    {
+                      "string": "userWebsite",
+                      "raw_string": "userWebsite"
+                    }
+                  ]
+                }
+              }
+            ]
+          },
+          "context": {
+            "edge": null,
+            "key": {
+              "range": "index.d2,0:0:0-0:11:11",
+              "key": {
+                "range": "index.d2,0:0:0-0:11:11",
+                "path": [
+                  {
+                    "unquoted_string": {
+                      "range": "index.d2,0:0:0-0:11:11",
+                      "value": [
+                        {
+                          "string": "userWebsite",
+                          "raw_string": "userWebsite"
+                        }
+                      ]
+                    }
+                  }
+                ]
+              },
+              "primary": {},
+              "value": {}
+            }
+          },
+          "due_to_glob": false,
+          "due_to_lazy_glob": false
+        }
+      ]
+    },
+    {
+      "name": "userMobile",
+      "composite": {
+        "fields": [
+          {
+            "name": "shape",
+            "primary": {
+              "value": {
+                "range": "x.d2,0:7:7-0:13:13",
+                "value": [
+                  {
+                    "string": "person",
+                    "raw_string": "person"
+                  }
+                ]
+              }
+            },
+            "references": [
+              {
+                "string": {
+                  "range": "x.d2,0:0:0-0:5:5",
+                  "value": [
+                    {
+                      "string": "shape",
+                      "raw_string": "shape"
+                    }
+                  ]
+                },
+                "key_path": {
+                  "range": "x.d2,0:0:0-0:5:5",
+                  "path": [
+                    {
+                      "unquoted_string": {
+                        "range": "x.d2,0:0:0-0:5:5",
+                        "value": [
+                          {
+                            "string": "shape",
+                            "raw_string": "shape"
+                          }
+                        ]
+                      }
+                    }
+                  ]
+                },
+                "context": {
+                  "edge": null,
+                  "key": {
+                    "range": "x.d2,0:0:0-0:13:13",
+                    "key": {
+                      "range": "x.d2,0:0:0-0:5:5",
+                      "path": [
+                        {
+                          "unquoted_string": {
+                            "range": "x.d2,0:0:0-0:5:5",
+                            "value": [
+                              {
+                                "string": "shape",
+                                "raw_string": "shape"
+                              }
+                            ]
+                          }
+                        }
+                      ]
+                    },
+                    "primary": {},
+                    "value": {
+                      "unquoted_string": {
+                        "range": "x.d2,0:7:7-0:13:13",
+                        "value": [
+                          {
+                            "string": "person",
+                            "raw_string": "person"
+                          }
+                        ]
+                      }
+                    }
+                  }
+                },
+                "due_to_glob": true,
+                "due_to_lazy_glob": false
+              }
+            ]
+          },
+          {
+            "name": "label",
+            "primary": {
+              "value": {
+                "range": "x.d2,1:7:21-1:11:25",
+                "value": [
+                  {
+                    "string": "meow",
+                    "raw_string": "meow"
+                  }
+                ]
+              }
+            },
+            "references": [
+              {
+                "string": {
+                  "range": "x.d2,1:0:14-1:5:19",
+                  "value": [
+                    {
+                      "string": "label",
+                      "raw_string": "label"
+                    }
+                  ]
+                },
+                "key_path": {
+                  "range": "x.d2,1:0:14-1:5:19",
+                  "path": [
+                    {
+                      "unquoted_string": {
+                        "range": "x.d2,1:0:14-1:5:19",
+                        "value": [
+                          {
+                            "string": "label",
+                            "raw_string": "label"
+                          }
+                        ]
+                      }
+                    }
+                  ]
+                },
+                "context": {
+                  "edge": null,
+                  "key": {
+                    "range": "x.d2,1:0:14-1:11:25",
+                    "key": {
+                      "range": "x.d2,1:0:14-1:5:19",
+                      "path": [
+                        {
+                          "unquoted_string": {
+                            "range": "x.d2,1:0:14-1:5:19",
+                            "value": [
+                              {
+                                "string": "label",
+                                "raw_string": "label"
+                              }
+                            ]
+                          }
+                        }
+                      ]
+                    },
+                    "primary": {},
+                    "value": {
+                      "unquoted_string": {
+                        "range": "x.d2,1:7:21-1:11:25",
+                        "value": [
+                          {
+                            "string": "meow",
+                            "raw_string": "meow"
+                          }
+                        ]
+                      }
+                    }
+                  }
+                },
+                "due_to_glob": true,
+                "due_to_lazy_glob": false
+              }
+            ]
+          }
+        ],
+        "edges": null
+      },
+      "references": [
+        {
+          "string": {
+            "range": "index.d2,1:0:12-1:10:22",
+            "value": [
+              {
+                "string": "userMobile",
+                "raw_string": "userMobile"
+              }
+            ]
+          },
+          "key_path": {
+            "range": "index.d2,1:0:12-1:10:22",
+            "path": [
+              {
+                "unquoted_string": {
+                  "range": "index.d2,1:0:12-1:10:22",
+                  "value": [
+                    {
+                      "string": "userMobile",
+                      "raw_string": "userMobile"
+                    }
+                  ]
+                }
+              }
+            ]
+          },
+          "context": {
+            "edge": null,
+            "key": {
+              "range": "index.d2,1:0:12-1:10:22",
+              "key": {
+                "range": "index.d2,1:0:12-1:10:22",
+                "path": [
+                  {
+                    "unquoted_string": {
+                      "range": "index.d2,1:0:12-1:10:22",
+                      "value": [
+                        {
+                          "string": "userMobile",
+                          "raw_string": "userMobile"
+                        }
+                      ]
+                    }
+                  }
+                ]
+              },
+              "primary": {},
+              "value": {}
+            }
+          },
+          "due_to_glob": false,
+          "due_to_lazy_glob": false
+        }
+      ]
+    }
+  ],
+  "edges": null
+}

